### DISCUSSION TITLE
preparing data from CP2K aimd output

### DISCUSSION THREAD

**Discussion by sli259** [deepmodeling/dpdata#374] [Q&A]
I am new to the deepMD and want to prepare a dp data from my aimd simulation with CP2K. I keep get this error about list index out of range. Can anyone give me some suggestions where I made the mistake?

```
d_outcar = dpdata.LabeledSystem('../cp2k_out/MnS-rs-pos-1.xyz', fmt = 'cp2k/aimd_output')

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In [3], line 4
      1 # dsys = dpdata.LabeledSystem('/Users/shili/Documents/ANL/projects/packages/deepmd-kit/examples/data_conv/OUTCAR')
      2 # d_poscar = dpdata.System('../cp2k_out/H2O-64-pos-1.xyz', fmt = 'cp2k/output')
      3 # d_outcar = dpdata.System('../trajectory/rs_traj.gro', fmt = 'gromacs/gro')
----> 4 d_outcar = dpdata.LabeledSystem('../cp2k_out/MnS-rs-pos-1.xyz', fmt = 'cp2k/aimd_output')
      5 # d_outcar = dpdata.LabeledSystem('../trajectory/rs_traj.gro', fmt = 'gromacs/gro')
      6 # d_outcar = dpdata.LabeledSystem('../cp2k_out/H2O.out', fmt = 'cp2k/output')
      7 # dsys = dpdata.LabeledSystem('gro', )

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/system.py:265, in System.__init__(self, file_name, fmt, type_map, begin, step, data, convergence_check, **kwargs)
    263 if file_name is None :
    264     return
--> 265 self.from_fmt(file_name, fmt, type_map=type_map, begin= begin, step=step, convergence_check=convergence_check, **kwargs)
    267 if type_map is not None:
    268     self.apply_type_map(type_map)

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/system.py:291, in System.from_fmt(self, file_name, fmt, **kwargs)
    289 if fmt == 'auto':
    290     fmt = os.path.basename(file_name).split('.')[-1].lower()
--> 291 return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/system.py:1073, in LabeledSystem.from_fmt_obj(self, fmtobj, file_name, **kwargs)
   1072 def from_fmt_obj(self, fmtobj, file_name, **kwargs):
-> 1073     data = fmtobj.from_labeled_system(file_name, **kwargs)
   1074     if data:
   1075         if isinstance(data, (list, tuple)):

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/plugins/cp2k.py:11, in CP2KAIMDOutputFormat.from_labeled_system(self, file_name, restart, **kwargs)
      9 def from_labeled_system(self, file_name, restart=False, **kwargs):
     10    # print(glob.glob("{}/*pos*.xyz".format(file_name)))
---> 11     xyz_file = sorted(glob.glob("{}/*pos*.xyz".format(file_name)))[0]
     12     log_file = sorted(glob.glob("{}/*.log".format(file_name)))[0]
     13     return tuple(Cp2kSystems(log_file, xyz_file, restart))

IndexError: list index out of range
```

The file is just an xyz contains several time frames. I am not sure where is the problem. it seems like the xyz_file in cp2k.py  didn't read any trajectories.

**Comment by njzjz** [comment]
It looks like the same error in #103.

**Reply by sli259** [reply]
Thanks for your suggestion, I have looked at the #103 and edited my code to read from the directory with xyz trajectory and output log file. Now there is no error of reading the files but somehow the information wasn't pass to the outcar. 

Here is my xyz and log file if you can take a look and let me know what I missed. Thank you!

[output.zip](https://github.com/deepmodeling/dpdata/files/10043426/output.zip)

```
d_outcar = dpdata.LabeledSystem('./cp2k_data/rs', fmt = 'cp2k/aimd_output')
print(d_outcar)

./cp2k_data/rs/pos.xyz ./cp2k_data/rs/rs.log

Data Summary
Labeled System
-------------------
Frame Numbers      : 0
Atom Numbers       : 0
Including Virials  : No
Element List       :
-------------------
```

**Comment by robinzyb** [comment]
Hi @sli259, the first argument of `dpdata.LabeledSystem` for cp2k aimd is a directory, not the name of pos files.

**Reply by sli259** [reply]
Thanks @robinzyb, I have adjusted it to the directory where I keep my xyz trajectory and the output log file. There is no error of reading these files but the information were not passed to the outcar. Do you any suggestion what I missed here?

**Reply by robinzyb** [reply]
Hi, I see. The version of cp2k is 9.1, which is not supported by dpdata.  I have developed a package specializing in processing data from cp2k. (see here [cp2kdata plugin for dpdata](https://github.com/robinzyb/cp2kdata#plug-in-for-dpdata)). just git clone and pip install it. Then you can use the new format "cp2kdata/aimd" for your outputs

**Reply by sli259** [reply]
Thanks @robinzyb, I installed the plugin and got the following error. Am I using the right fmt? I tried cp2kdata/aimd but it didn't recognize the format. 


```
d_outcar = dpdata.LabeledSystem('./cp2k_data/rs', cp2k_output_name = 'rs.log', fmt = 'cp2kdata/md')

--- You are parsing data using package Cp2kData ---
You are reading cell information from ./cp2k_data/rs/rs.log
Obtian Structures From ./cp2k_data/rs/pos.xyz
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Cell In [4], line 8
      1 # dsys = dpdata.LabeledSystem('/Users/shili/Documents/ANL/projects/packages/deepmd-kit/examples/data_conv/OUTCAR')
      2 # d_poscar = dpdata.System('../cp2k_out/H2O-64-pos-1.xyz', fmt = 'cp2k/output')
      3 # d_outcar = dpdata.System('../trajectory/rs_traj.gro', fmt = 'gromacs/gro')
   (...)
      6 # cp2kmd_dir = './cp2k_data/rs'
      7 # cp2kmd_output_name = 'rs.log'
----> 8 d_outcar = dpdata.LabeledSystem('./cp2k_data/rs', cp2k_output_name = 'rs.log', fmt = 'cp2kdata/md')
     10 # d_outcar = dpdata.LabeledSystem('./cp2k_data/rs', fmt = 'cp2kdata/md')
     11 # print(d_outcar)
     12 # d_outcar = dpdata.LabeledSystem('../trajectory/rs_traj.gro', fmt = 'gromacs/gro')
     13 # d_outcar = dpdata.LabeledSystem('../cp2k_out/H2O.out', fmt = 'cp2k/output')
     14 # dsys = dpdata.LabeledSystem('gro', )

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/system.py:265, in System.__init__(self, file_name, fmt, type_map, begin, step, data, convergence_check, **kwargs)
    263 if file_name is None :
    264     return
--> 265 self.from_fmt(file_name, fmt, type_map=type_map, begin= begin, step=step, convergence_check=convergence_check, **kwargs)
    267 if type_map is not None:
    268     self.apply_type_map(type_map)

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/system.py:291, in System.from_fmt(self, file_name, fmt, **kwargs)
    289 if fmt == 'auto':
    290     fmt = os.path.basename(file_name).split('.')[-1].lower()
--> 291 return self.from_fmt_obj(load_format(fmt), file_name, **kwargs)

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/dpdata/system.py:1073, in LabeledSystem.from_fmt_obj(self, fmtobj, file_name, **kwargs)
   1072 def from_fmt_obj(self, fmtobj, file_name, **kwargs):
-> 1073     data = fmtobj.from_labeled_system(file_name, **kwargs)
   1074     if data:
   1075         if isinstance(data, (list, tuple)):

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/cp2kdata/dpdata_plugin.py:71, in CP2KMDFormat.from_labeled_system(self, file_name, **kwargs)
     68 # -- start parsing -- 
     69 print(WRAPPER)
---> 71 cp2kmd = Cp2kOutput(output_file=cp2k_output_name, run_type="MD", path_prefix=path_prefix)
     73 num_frames = cp2kmd.get_num_frames()
     75 chemical_symbols = get_chemical_symbols_from_cp2kdata(
     76     cp2koutput=cp2kmd, 
     77     true_symbols=true_symbols
     78     )

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/cp2kdata/output.py:100, in Cp2kOutput.__init__(self, output_file, run_type, path_prefix)
     92 Parse_Run_Type = {
     93     "ENERGY_FORCE": self.parse_energy_force,
     94     "GEO_OPT": self.parse_geo_opt,
     95     "CELL_OPT": self.parse_cell_opt,
     96     "MD": self.parse_md
     97 }
     99 # call respective parser for run types
--> 100 Parse_Run_Type[self.GlobalInfo.run_type]()

File ~/miniconda3/envs/deepmd/lib/python3.8/site-packages/cp2kdata/output.py:390, in Cp2kOutput.parse_md(self)
    387 else:
    388     self.stress_tensor_list = None
--> 390 self.num_frames = len(self.energies_list)

AttributeError: 'Cp2kOutput' object has no attribute 'energies_list'

```

**Reply by robinzyb** [reply]
seems cp2kdata automatically find 'pos.xyz' as trajectory file. Is this file the cp2k trajectory file ? if not, you can rename it  or just move it out the directory. The code matches the trajectory file using `*pos*`

**Reply by sli259** [reply]
This is the trajectory file, I only have the log file and the pos.xyz in this directory. I have attached these files here, would you mind to run the code and see if you got the same error? Thank you so much!
[cp2k.zip](https://github.com/deepmodeling/dpdata/files/10068244/cp2k.zip)

**Reply by robinzyb** [reply]
hi, I found the problem. you can print FORCE and TRAJECTORY using [CP2K_INPUT](https://manual.cp2k.org/cp2k-8_2-branch/CP2K_INPUT.html) / [MOTION](https://manual.cp2k.org/cp2k-8_2-branch/CP2K_INPUT/MOTION.html) / [PRINT](https://manual.cp2k.org/cp2k-8_2-branch/CP2K_INPUT/MOTION/PRINT.html), which will write `xx-frc-1.xyz` and `xx-pos-1.xyz` files. This is a more efficient way to get force information/ also stress. parse force from output(log file) are supported for energy_force calculation, and can be implemented in md parser with little effort if you would like to wait me two days.

**Reply by robinzyb** [reply]
By the way, I noticed the SCF are not converged in output, which may obtain incorrect forces.

**Reply by sli259** [reply]
Sure, thank you for your help!

**Reply by robinzyb** [reply]
Hi, if possible, you can move the problem to the issue in cp2kdata repository. I would like to develop a more user friendly and useful python tool to quickly extract information from various cp2k outputs

**Reply by sli259** [reply]
Yes, thanks for the notice. I only run this for a short time to test if I can generate the dp data. Can you give me some suggestions what I can adjust to my input to use your parser at this moment?

**Reply by robinzyb** [reply]
YES, activate the print/trajectory print/forces section as I mentioned.  set EACH/MD as 1

**Reply by sli259** [reply]
> Hi, if possible, you can move the problem to the issue in cp2kdata repository. I would like to develop a more user friendly and useful python tool to quickly extract information from various cp2k outputs

Hi I am not sure how to transfer this to the issue in cp2kdata, it looks like I can only transfer the discussion to repository owned by deepmodeling.

**Comment by yetnobe** [comment]
Hi I got a similar error.
When I run the output file you provided, there are no errors. I would appreciate it if you could provide me with your "inp" file.

**Reply by robinzyb** [reply]
have you tried cp2kdata plugin?

**Reply by yetnobe** [reply]
oh thanks!! I will check with cp2k plugin.
I am a noob to cp2k.
I generate the "inp" file with "Multiwfn" and then manully modify the print conents.

[inp.txt](https://github.com/deepmodeling/dpdata/files/13386949/inp.txt)
